libdevcore/SHA3.cpp

```diff
@@ -188,55 +174,13 @@ static inline int hash(uint8_t* out, size_t outlen,
 	return 0;
 }
 
-/*** Helper macros to define SHA3 and SHAKE instances. ***/
-#define defshake(bits)                                            \
-  int shake##bits(uint8_t* out, size_t outlen,                    \
-				  const uint8_t* in, size_t inlen) {              \
-	return hash(out, outlen, in, inlen, 200 - (bits / 4), 0x1f);  \
-  }
-#define defsha3(bits)                                             \
-  int sha3_##bits(uint8_t* out, size_t outlen,                    \
-				  const uint8_t* in, size_t inlen) {              \
-	if (outlen > (bits/8)) {                                      \
-	  return -1;                                                  \
-	}                                                             \
-	return hash(out, outlen, in, inlen, 200 - (bits / 4), 0x06);  \
-  }
-#define defkeccak(bits)                                             \
-  int keccak##bits(uint8_t* out, size_t outlen,                    \
-				  const uint8_t* in, size_t inlen) {              \
-	if (outlen > (bits/8)) {                                      \
-	  return -1;                                                  \
-	}                                                             \
-	return hash(out, outlen, in, inlen, 200 - (bits / 4), 0x01);  \
-  }
-
-/*** FIPS202 SHAKE VOFs ***/
-defshake(128)
-defshake(256)
-
-/*** FIPS202 SHA3 FOFs ***/
-defsha3(224)
-defsha3(256)
-defsha3(384)
-defsha3(512)
-
-/*** KECCAK FOFs ***/
-defkeccak(224)
-defkeccak(256)
-defkeccak(384)
-defkeccak(512)
-
 }
 
-bool keccak256(bytesConstRef _input, bytesRef o_output)
+h256 keccak256(bytesConstRef _input)
 {
-	// FIXME: What with unaligned memory?
-	if (o_output.size() != 32)
-		return false;
-	keccak::keccak256(o_output.data(), 32, _input.data(), _input.size());
-//	keccak::keccak(ret.data(), 32, (uint64_t const*)_input.data(), _input.size());
-	return true;
+	h256 output;
+	hash(output.data(), output.size, _input.data(), _input.size(), 200 - (256 / 4), 0x01);
```

Comment:
I think it is worth leaving at least a comment explaining the parameters.

The 0x01 is the specific padding for keccak (sha3 uses 0x06) and the way the round size (or window or whatever it was) is calculated.

Reply:
hm, the params have names in the function "hash", not sure we still need to comment here.

Reply:
Well the rate rate calculation is not well explained with only having constants there. At least a comment for that could make sense.